hibernate-reactive-core/src/test/java/org/hibernate/reactive/CompletionStagesTest.java

```diff
@@ -0,0 +1,181 @@
+/* Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright: Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.reactive;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.IntStream;
+
+import org.hibernate.reactive.util.impl.CompletionStages;
+
+import static org.hibernate.reactive.util.impl.CompletionStages.total;
+import static org.hibernate.reactive.util.impl.CompletionStages.loop;
+import static org.hibernate.reactive.util.impl.CompletionStages.loopWithoutTrampoline;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import io.vertx.ext.unit.Async;
+import io.vertx.ext.unit.TestContext;
+import io.vertx.ext.unit.junit.RunTestOnContext;
+import io.vertx.ext.unit.junit.VertxUnitRunner;
+import org.assertj.core.api.Assertions;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@RunWith(VertxUnitRunner.class)
+public class CompletionStagesTest {
+
+	@Rule
+	public RunTestOnContext vertxContextRule = new RunTestOnContext();
+
+	protected static void test(TestContext context, CompletionStage<?> cs) {
+		// this will be added to TestContext in the next vert.x release
+		Async async = context.async();
+		cs.whenComplete( (res, err) -> {
+			if ( err != null ) {
+				context.fail( err );
+			}
+			else {
+				async.complete();
+			}
+		} );
+	}
+
+	@Test
+	public void testTotalWithIntegers(TestContext context) {
+		int startInt = 0;
+		int endInt = 11;
+
+		// create a value of 'index + 2' and add it to the total
+		test(   context,
+			    total( startInt, endInt, index -> CompletionStages.completedFuture( index + 2 ) )
+					  .thenAccept( total -> Assertions.assertThat( total ).isEqualTo( 77 ) )
+		);
+	}
+
+	@Test
+	public void testTotalWithIterator(TestContext context) {
+		List<Object> entries = Arrays.asList( 1, 2, 3, 4, 5 );
+		int incrementPerEntry = 3;
+
+		test(   context,
+			    total( entries.iterator(), increment -> CompletionStages.completedFuture( incrementPerEntry ) )
```

Comment:
My point is that this test only check the final total but doesn't check if the elements are looped in the right order.
I think we should check both. This might be important for us because if we run several updates queries (and get the number of rows changed), we want 
to make sure that they run in the right order.

Reply:
I mean something like this:
```
entry -> CompletionStages.voidFuture()
                  .thenAccept(v -> looped.add(entry))
                  .thenApply(v -> incrementPerEntry)
```

And then you check looped as usual.
I wrote it without testing it, it might need some tweaking.

Reply:
that worked.  I added the check on entries to the other `total()` method too.

Reply:
done